From fb75ab74e75afc2588f17948f872b02fa852c6f8 Mon Sep 17 00:00:00 2001
From: Vincent Becker <vincentx.becker@intel.com>
Date: Mon, 12 Nov 2012 10:07:56 +0100
Subject: [PATCH 18/49] [PORT FROM MAIN] [AUDIO_FM] Add forced usage support
 for FM radio in audio policy

BZ: 67617

Add the support of forced usage FOR_FM_RADIO, specific to
the FM radio strategy. It mainly concerns when the user selects
the speaker output on the FM application..

Change-Id: I14db2e41c1a09a5c414295e3650a74d980da5820
Origin-Change-Id: I12b4f04d9b9c1d80f0bffc832d9fea587b14a4d3
Signed-off-by: Vincent Becker <vincentx.becker@intel.com>
Reviewed-on: http://android.intel.com:8080/78834
Tested-by: Barthes, FabienX <fabienx.barthes@intel.com>
Reviewed-by: Barthes, FabienX <fabienx.barthes@intel.com>
Reviewed-by: cactus <cactus@intel.com>
Tested-by: cactus <cactus@intel.com>
---
 audio/AudioPolicyManagerBase.cpp            |   12 +++++++++---
 include/hardware_legacy/AudioSystemLegacy.h |    5 +++--
 2 files changed, 12 insertions(+), 5 deletions(-)

diff --git a/audio/AudioPolicyManagerBase.cpp b/audio/AudioPolicyManagerBase.cpp
index ddbc3a1..53ae32f 100644
--- a/audio/AudioPolicyManagerBase.cpp
+++ b/audio/AudioPolicyManagerBase.cpp
@@ -472,9 +472,15 @@ void AudioPolicyManagerBase::setForceUse(AudioSystem::force_use usage, AudioSyst
             config != AudioSystem::FORCE_SYSTEM_ENFORCED) {
             ALOGW("setForceUse() invalid config %d for FOR_SYSTEM", config);
         }
+    case AudioSystem::FOR_FM_RADIO:
+        if (config != AudioSystem::FORCE_NONE && config != AudioSystem::FORCE_SPEAKER &&
+            config != AudioSystem::FORCE_HEADPHONES && config != AudioSystem::FORCE_WIRED_ACCESSORY) {
+            ALOGW("setForceUse() invalid config %d for FOR_FM_RADIO", config);
+        }
         forceVolumeReeval = true;
         mForceUse[usage] = config;
         break;
+
     default:
         ALOGW("setForceUse() invalid usage %d", usage);
         break;
@@ -2326,7 +2332,7 @@ audio_devices_t AudioPolicyManagerBase::getDeviceForStrategy(routing_strategy st
 
         uint32_t device2 = 0;
 
-        if(mForceUse[AudioSystem::FOR_MEDIA] == AudioSystem::FORCE_SPEAKER) {
+        if(mForceUse[AudioSystem::FOR_FM_RADIO] == AudioSystem::FORCE_SPEAKER) {
             LOGV("getDeviceForStrategy():Force use of SPEAKER for STRATEGY_FM_RADIO.");
             device2 = mAvailableOutputDevices & AudioSystem::DEVICE_OUT_SPEAKER;
         }
@@ -2563,11 +2569,11 @@ audio_devices_t AudioPolicyManagerBase::getDeviceForInputSource(int inputSource)
             device = AUDIO_DEVICE_IN_VOICE_CALL;
         }
         break;
-    case AUDIO_SOURCE_REMOTE_SUBMIX:
+   case AUDIO_SOURCE_REMOTE_SUBMIX:
         if (mAvailableInputDevices & AUDIO_DEVICE_IN_REMOTE_SUBMIX) {
             device = AUDIO_DEVICE_IN_REMOTE_SUBMIX;
         }
-        break;
+        break; 
     case AUDIO_SOURCE_FM:
         if (mAvailableInputDevices & AudioSystem::DEVICE_IN_FM_RECORD) {
             device = AudioSystem::DEVICE_IN_FM_RECORD;
diff --git a/include/hardware_legacy/AudioSystemLegacy.h b/include/hardware_legacy/AudioSystemLegacy.h
index 6a69be4..996e9cd 100644
--- a/include/hardware_legacy/AudioSystemLegacy.h
+++ b/include/hardware_legacy/AudioSystemLegacy.h
@@ -61,7 +61,7 @@ enum audio_source {
     AUDIO_SOURCE_CAMCORDER = 5,
     AUDIO_SOURCE_VOICE_RECOGNITION = 6,
     AUDIO_SOURCE_VOICE_COMMUNICATION = 7,
-    AUDIO_SOURCE_FM = 8,
+    AUDIO_SOURCE_FM = 9,
     AUDIO_SOURCE_MAX = AUDIO_SOURCE_FM,
     AUDIO_SOURCE_LIST_END  // must be last - used to validate audio source type
 };
@@ -275,7 +275,7 @@ public:
 
         DEVICE_IN_ALL = (DEVICE_IN_COMMUNICATION | DEVICE_IN_AMBIENT | DEVICE_IN_BUILTIN_MIC |
                 DEVICE_IN_BLUETOOTH_SCO_HEADSET | DEVICE_IN_WIRED_HEADSET | DEVICE_IN_AUX_DIGITAL |
-                DEVICE_IN_VOICE_CALL | DEVICE_IN_BACK_MIC | DEVICE_IN_FM_RECORD | DEVICE_IN_DEFAULT)
+                DEVICE_IN_VOICE_CALL | DEVICE_IN_BACK_MIC | DEVICE_IN_FM_RECORD | DEVICE_IN_DEFAULT),
 
         //devices not supported for codec offload
         DEVICE_OUT_NON_OFFLOAD = (DEVICE_OUT_ALL & ~(DEVICE_OUT_SPEAKER |
@@ -317,6 +317,7 @@ public:
         FOR_RECORD,
         FOR_DOCK,
         FOR_SYSTEM,
+        FOR_FM_RADIO,
         NUM_FORCE_USE
     };
 
-- 
1.7.9

